### PR TITLE
Create a simple Dockerfile for the server

### PR DESCRIPTION
**Pull request by balandi1**
- Initial Dockerfile creation for the server

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,3 @@
-# image-builder
\ No newline at end of file
+# Image Builder
+An application that accepts desired configuration and creates Dockerfile.
```
Comment (pradykaushik) [inline README.md]: The application should create an image. Not just a dockerfile.

Comment (pradykaushik) [reply README.md]: Also describe what this "configuration" is.

File: config.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+serverConfig:
```
Comment (pradykaushik) [inline config.yaml]: I feel that configuration pertaining to `code-runner` should be in a separate yaml file, say `code-runner.yaml`.
Keep in mind that this application is intended to be used by professors/teaching staff to build an image that has the environment to run the assignments.

The assignment runtime environment information should be part of another yaml file, say `assignment-runtime-env.yaml`.

Typically, the `code-runner.yaml` will be left untouched. Students can update that if they want to change the port on which the code-runner runs.

File: dockerConfig/createDockerfile.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+)
+
+// Generates a dockerfile instructions as a string
+func render(config dockerfileData) string {
```
Comment (pradykaushik) [inline dockerConfig/createDockerfile.go]: what is the rationale behind naming the function `render`?

File: dockerConfig/createDockerfile.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	result := fmt.Sprintf("FROM %s\n", config.ServerConfig.From)
+
+	for goEnv, value := range config.ServerConfig.Env {
+		result += fmt.Sprintf("ENV %s=%s\n", goEnv, value)
+	}
+
+	result += fmt.Sprintf("COPY %s %s\n", config.ServerConfig.Copy["baseDir"], config.ServerConfig.Copy["destDir"])
```
Comment (pradykaushik) [inline dockerConfig/createDockerfile.go]: This is not a good way of creating the dockerfile.
You should create a template and then build each component of the dockerfile separately. Finally, the template, along with the values for each field should be serialized to create a dockerfile.

File: dockerConfig/createDockerfile.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+	result += fmt.Sprintf("EXPOSE %s\n", config.ServerConfig.ServerPort)
+
+	result += fmt.Sprintf("CMD [\"%s\"]\n", config.ServerConfig.FinalCmd)
+
+	fmt.Println(result)
+	return result
+}
+
+// creates and writes the rendered string into dockerfile
```
Comment (pradykaushik) [inline dockerConfig/createDockerfile.go]: Follow golang format for writing comments - https://blog.golang.org/godoc.

File: dockerConfig/createDockerfile.go (hunk 4 above)
Comment (pradykaushik) [reply dockerConfig/createDockerfile.go]: Take a look at this - https://blog.dockbit.com/templating-your-dockerfile-like-a-boss-2a84a67d28e9

File: dockerConfig/createDockerfile.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+func WriteDockerfile(config dockerfileData) error {
+	result := render(config)
+	file, err := os.Create("Dockerfile")
+	if err != nil {
+		return err
```
Comment (pradykaushik) [inline dockerConfig/createDockerfile.go]: we have a resource leak at this stage.

File: dockerConfig/dockerDataTemplate.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package dockerConfig
+
+import (
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"log"
+)
+
+type dockerfileData struct {
```
Comment (pradykaushik) [inline dockerConfig/dockerDataTemplate.go]: Why compose `ServerConfig` in `dockerfileData`?

File: dockerConfig/dockerDataTemplate.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	yamlFile, err := ioutil.ReadFile(configFilename)
+	if err != nil {
+		log.Println(err)
+		return nil, err
+	}
+
+	config := &dockerfileData{}
+	err = yaml.Unmarshal(yamlFile, config)
+	if err != nil {
+		log.Printf("Error in unmarshalling yaml: %v", err)
```
Comment (pradykaushik) [inline dockerConfig/dockerDataTemplate.go]: error messages are always lowercase without full-stops.

File: imageBuilder.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package main
+
+import (
+	"assignment-exec/image-builder/dockerConfig"
```
Comment (pradykaushik) [inline imageBuilder.go]: let's not use camelcasing for package names. It's against golang conventions - https://blog.golang.org/package-names

File: imageBuilder.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+	"fmt"
+	"log"
+)
+
+func main() {
+	fmt.Println("Creating Dockerfile...")
+
+	// read the yaml config
```
Comment (pradykaushik) [inline imageBuilder.go]:
```go
// Unmarshal the yaml configuration file.
```

**Review by pradykaushik — CHANGES_REQUESTED** [review]
For code comments follow the conventions here - https://blog.golang.org/godoc.
At the very basic, each comment should start with a capital letter and end with a full stop.

File: dockerConfig/createDockerfile.go (hunk 4 above)
Comment (balandi1) [reply dockerConfig/createDockerfile.go]: Yes. Working on it.

File: dockerConfig/createDockerfile.go (hunk 5 above)
Comment (balandi1) [reply dockerConfig/createDockerfile.go]: Ok

File: dockerConfig/dockerDataTemplate.go (hunk 7 above)
Comment (balandi1) [reply dockerConfig/dockerDataTemplate.go]: I was thinking of having server config and assignment config in one struct. But as said by you, these two will be separately handled

File: dockerConfig/dockerDataTemplate.go (hunk 8 above)
Comment (balandi1) [reply dockerConfig/dockerDataTemplate.go]: okay

File: imageBuilder.go (hunk 9 above)
Comment (balandi1) [reply imageBuilder.go]: okay

File: config.yaml (hunk 2 above)
Comment (balandi1) [reply config.yaml]: Okay. Got it

File: dockerConfig/createDockerfile.go (hunk 3 above)
Comment (balandi1) [reply dockerConfig/createDockerfile.go]: Renamed it

File: dockerConfig/createDockerfile.go (hunk 6 above)
Comment (balandi1) [reply dockerConfig/createDockerfile.go]: Are you mentioning about `file`

File: imageBuilder.go (hunk 10 above)
Comment (balandi1) [reply imageBuilder.go]: Done

File: README.md (hunk 1 above)
Comment (balandi1) [reply README.md]: Done

File: dockerConfig/createDockerfile.go (hunk 6 above)
Comment (pradykaushik) [reply dockerConfig/createDockerfile.go]: Yes. File is not closed at line 37.

File: README.md [hunk 11]
```diff
@@ -1 +1,39 @@
-# image-builder
\ No newline at end of file
+# Image Builder
+Image builder is an application to build docker image for running assignments. The image is built using user specific configurations.
+
+## Configurations
+There are two types of configurations provided - Code runner and Assignment Environment.
+
+### Code Runner Configurations
+- These configurations are required for running the code-runner server.
```
Comment (pradykaushik) [inline README.md]: Change `code-runner` to `[code-runner](https://github.com/assignment-exec/code-runner)`.

File: README.md [hunk 12]
```diff
@@ -1 +1,39 @@
...
+- Code runner configurations are specified in yaml format and include instructions that are needed in Dockerfile for running the server.
+Following is a sample of the configuration yaml
+```commandline
+stages:
```
Comment (pradykaushik) [inline README.md:12]: What are the stages?
I feel "stages" doesn't make sense here. Why have another level of indirection when you can just have the top level as `serverConfig`?

File: code-runner.yaml [hunk 13]
```diff
@@ -0,0 +1,20 @@
+stages:
```
Comment (pradykaushik) [inline code-runner.yaml:1]: refer to "stages" comment made eariler.

File: configurations/configurations_test.go [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package configurations
+
+import (
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+var expectedGenericOutput = `FROM golang:latest As golang
```
Comment (pradykaushik) [inline configurations/configurations_test.go]: why "As golang"?

File: configurations/createDockerfile.go [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+	data, err := newDockerFileDataFromYamlFile("code-runner.yaml")
+	if err != nil {
+		return err
+	}
+
+	tmpl := newDockerfileTemplate(data)
+
+	file, err := os.Create("Dockerfile")
+	if err != nil {
+		return err
```
Comment (pradykaushik) [inline configurations/createDockerfile.go]: resource leak here. The defer function should be placed before any return statement.
In this case, you should place right after line 17.

File: configurations/dockerDataTemplate.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+	"fmt"
+	"strings"
+)
+
+type instruction interface {
+	WriteInstruction() string
+}
+
+type dockerfileData struct {
+	Stages []stage
```
Comment (pradykaushik) [inline configurations/dockerDataTemplate.go:13]: If we get rid of "stages" (as discussed earlier), then this struct is no longer required.

File: configurations/dockerDataTemplate.go [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+// Decodes the yaml data and gives the stage instance having all the dockerfile instructions.
+func (s *stage) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var data []interface{}
+	var result []instruction
+	err := unmarshal(&data)
+	if err != nil {
+		return err
+	}
+	*s = append(result, getInstructions(data)...)
```
Comment (pradykaushik) [inline configurations/dockerDataTemplate.go]: why not just `*s = getInstructions(data)`? Do we need `result`?

File: README.md (hunk 12 above)
Comment (pradykaushik) [reply README.md:12]: If it's going to be too much of a change, then we can take it up in another PR. Just make a note of it though.

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (balandi1) [reply README.md]: Done

File: configurations/configurations_test.go (hunk 14 above)
Comment (balandi1) [reply configurations/configurations_test.go]: `As` is not needed here. I was just testing whether all the fields of template are working or no.

File: configurations/createDockerfile.go (hunk 15 above)
Comment (balandi1) [reply configurations/createDockerfile.go]: Done

File: configurations/dockerDataTemplate.go (hunk 16 above)
Comment (balandi1) [reply configurations/dockerDataTemplate.go:13]: Yes agreed

File: README.md (hunk 12 above)
Comment (balandi1) [reply README.md:12]: I had kept stages, just to keep my template structure generic. If at all there are multiple stages required in the dockerfile, then my template structure should be able to easily adopt it by simply adding another stage data into yaml.

File: code-runner.yaml (hunk 13 above)
Comment (balandi1) [reply code-runner.yaml:1]: Okay

File: configurations/dockerDataTemplate.go (hunk 17 above)
Comment (balandi1) [reply configurations/dockerDataTemplate.go]: Yes I agree. `result` has no significance here. In the initial code build, I was modifying the result and then assigning it to `*s`. Later I forgot to get rid of `result`.

File: README.md (hunk 12 above)
Comment (balandi1) [reply README.md:12]: But as per your suggestion, I will remove the `stages` and do the rest of required changes in next PR.

**Review by pradykaushik — APPROVED** [review]

**Pull request merged**